### ISSUE TITLE
stepper fails for unbound id in test position of check-expect

### ISSUE THREAD

**Issue by jbclements** [racket/htdp#153]
To reproduce:

1) Set language level to Beginner
2) put this program in defns window:

(check-expect (f (list 4))
              (list 4 13))
(check-expect (f (list 4))
              (list 7 13))

Click Step.

See this error:

recon-source: no matching clause for syntax: (quote-syntax f)

Looks like in this case the reconstructed source contains a quote-syntax. 

This bug has been present since at least Racket 7.4.

**Comment by shhyou** [comment]
The test engine recently fixed the source location of syntax errors in check-expects (#152). Is it related or is the bug exposed by that?

**Comment by jbclements** [comment]
perhaps indirectly. I think the issue here is that the reconstructor is just blindsided by the appearance of a quote-syntax, which suggests that it has to do with the expansion of the test forms. Could be related, though. Programming is complicated....

**Comment by mikesperber** [comment]
> perhaps indirectly. I think the issue here is that the reconstructor is just blindsided by the appearance of a quote-syntax, which suggests that it has to do with the expansion of the test forms. 

The expansion of tests forms certainly did change, but there's no (obvious) quote-syntax sitting there. (And the macros stepper refuses to work in the presence of undefined variables, AFAICS.)

The relevant macro is here, if that helps:

https://github.com/racket/htdp/blob/fc6c2c264179b14fa6cb9ecb1ccdfd69df03a757/htdp-lib/test-engine/syntax.rkt#L29

**Comment by mikesperber** [comment]
So it's this line:

https://github.com/racket/htdp/blob/29e1c78b43769a778e754b18625b6125f337cdb8/htdp-lib/test-engine/syntax.rkt#L40

... which was recently discussed elsewhere.

Changing it to:

```racket
  (define test-expr-checked-for-syntax-error test-expr)
```

fixes the problem, makes the stepper refuse to start, with an appropriate error message.

**Comment by rfindler** [comment]
Does removing this line generally make the stepper work in a way we like with `#lang`-based `check-expect` or is there more to sort out here?

**Comment by mikesperber** [comment]
This problem is for both menu-based and `#lang`- so the bug fix is for both. AFAIK, the stepper has generally been fine with `#lang`.

**Comment by jbclements** [comment]
I meant to take a closer look at this ... but I haven't had time. @mikesperber , isn't this change simply removing a call to `convert-compile-time-error` ? I have no idea what this call was supposed to do, but is it safe to just remove it? I think you know way more about the test-engine here, so I guess I'm just asking you to explain the code to me, sorry...

Second question: AFAIK the #lang-based stepper directs the output of check-expect failures to DrRacket's stdout, which appears on the user's terminal... if they started DrRacket using a terminal, which for students is almost certainly not the case. In other words, if check-expect failures are now displayed correctly for #lang-based languages, this would be a big step forward! Not that I expect that, I would be very surprised, I think I'm just trying to clarify what robby said.

**Comment by mikesperber** [comment]
> Second question: AFAIK the #lang-based stepper directs the output of check-expect failures to DrRacket's stdout, which appears on the user's terminal... 

Whoopsie, that's a bug - wanna create a separate issue?

**Comment by mikesperber** [comment]
> @mikesperber , isn't this change simply removing a call to `convert-compile-time-error` ? I have no idea what this call was supposed to do, but is it safe to just remove it? I think you know way more about the test-engine here, so I guess I'm just asking you to explain the code to me, sorry...

@jbclements Yes, that's what the change is. Nobody remembers why the `convert-compile-time-error` was there in the first place, and we considered removing it for other reasons recently. So we'll run with it for now and see if it causes any problems.

**Comment by jbclements** [comment]
It looks to me like this use of `convert-compile-time-error` was inserted in 7143cd57738a88223f0285c63c1e  by @mfelleisen , cleaning up an existing use of local-expand that the comments suggest was intended to deal with arity errors. Could removing this cause arity errors in the test expression to be handled badly?

**Comment by samth** [comment]
Right now, this program runs two tests and reports a failure in the first one in BSL:

```
(check-expect (1 + 2) 3)
(check-expect 3 (+ 1 2))
```

Unfortunately this is still a static error that prevents tests from running:

```
(check-expect 3 (1 + 2))
```

I think we should make _more_ tests run in the face of syntax errors.

**Comment by jbclements** [comment]
Ah! Okay, so this is the age-old "can we provide more useful information if we don't just halt immediately on the first compilation error" goal. I ... don't really have a dog in this fight, and I think the stepper can be made to work either way, with a certain amount of work (specifically, it looks like judicious uses of "stepper-skip-completely" are the solution if we decide to stick with or extend the current direction; that is, assuming the wrapped pieces of code aren't actually going to be executed at all.

**Comment by mfelleisen** [comment]
Sorry I busy until now .. and I pointed out that we needed to re-run the tests in “htdp-tests” before we push this through — in an off-line email thread that you can’t see. 


(Now that @jbclements <https://github.com/jbclements> is pointing out this commit, it starts ringing a bell or even two.)

**Comment by mflatt** [comment]
Should the commit be reverted for now? I think the Northwestern snapshot is going to fail, since it creates some distributions with "main-distribution-test".

**Comment by mfelleisen** [comment]
Let’s figure out what’s really going on.  If Robby or Mike know,  let’s stick with it. If not, let’s revert and let’s study the failing test case.

**Comment by mfelleisen** [comment]
Following up to my own message. From the parallel off-line mail thread with just Mike: 



> On Sep 10, 2024, at 3:23 AM, Michael Sperber ***@***.***> wrote:
> 
> On Mon, Sep 09 2024, ***@***.*** wrote:
> 
>> Did none of tests in test-engine/ fail? 
> 
> Ah sorry, I got confused by all those little branches and PRs I made
> over the weekend, and must have run the tests without the commit in it:
> Some do fail (for reasons I don't understand) - I've pushed a fix.
> 
> (The problem is not the call to convert-compile-time-error, but how the
> code is spliced together.)
> 
> -- 
> Regards,
> Mike


Looks like we’re okay to go with this commit.

**Comment by mfelleisen** [comment]
@mikesperber I do not understand " (The problem is not the call to convert-compile-time-error, but how the
 code is spliced together.)".  

@mflatt (and everyone else I guess) -- I took a look at the code and the failing test case that came with Mike's change. I am beginning to recall why I shifted the mistake to compile time.  The rationale is coming back to me. Someone sent in an email to complain that in a file such as this: 

```
(define (f x) x)
(check-expect (f 1 2) 1)
(check-expect (f) 2)
```

only the first error is signaled to students and, after fixing it, they get a second one. Since BSL defines functions as syntax, this form of arity checking could be done at compile time so I inserted the shift (and then found someone had written a robust version and re-used this one).  

(The test file mentions `check-satisfied` and that makes me think DVH requested such a change because he was teaching our F I at Maryland at the time, and I made several changes in response.) 

The sad thing is that I didn't comment the added tests properly and/or leave a better rationale in the commit message to say why we're doing this. 

We have two options: 

1. roll back Mike's commit(s)
2. modify the test to focus on just the plural/singular issue of the error messages (which was a _secondary_ concern). 

What speaks for 1: The design of BSL makes it much more static than Racket.  So why not check for such properties early. 
What speaks for 2: If we are still thinking of BSL as a step toward Racket where you get one error at a time, we should stick with Mike's change and adjust the test. 

My own sense is that 1 is something good for @rfindler 's soul and probably good for students' spirit of overcoming early hurdles in a course like F I. At the same time, uniformity and fixing the stepper speak for 2. 

Opinions? 

(And again sorry for the delay. Sw Dev kept me busy.)

**Comment by samth** [comment]
In my post https://github.com/racket/htdp/issues/153#issuecomment-2338712890 I gave two examples of behavior. Are you proposing to change the current (ie 8.14) behavior of either of those two examples? In which way?

**Comment by mflatt** [comment]
I'll revert d70110566adc9e217fddd80c62af3091107f660b for now. It broke the Utah snapshots, too, as well as daily the build plots, and it's showing up as an error for anyone who builds from a Git checkout.

**Comment by shhyou** [comment]
@mikesperber In our class, it is important that the `expression` position of `check-expect` do **NOT** raise unbound-identifier error at expansion time. We commonly provide _exemplar_ `check-expect`s in our start code for the purpose of explaining the problem specification like this:

```racket
#lang htdp/isl+

;; add2 : number -> number
;; Adds 2 to the input.
;;
;; Example:
(check-expect (add2 10) 12)
```

I hope such code would not err.

**Comment by mikesperber** [comment]
Re-opening, as the patch got reverted.
